### PR TITLE
[TML] Changing return variable assignement in timestepping

### PR DESCRIPTION
**Pull request by jbathmann**
addressing Issue #136: some reformulation of variable assignement in order to avoid confusions

### CODE REVIEW

File: TimeLoopLib/SimpleTimeLoop/SimpleLoop.py [hunk 1]
```diff
@@ -27,22 +27,21 @@ def __init__(self, args):
...
+        self.t_1 = self.t_start
 
     def getUpdateBools(self):
-        update_ag = True if self.step_counter % self.n_step_ag == 0 else False
-        update_bg = True if self.step_counter % self.n_step_bg == 0 else False
-        return update_ag, update_bg
+        self.update_ag = True if self.step_counter % self.n_step_ag == 0 else False
+        self.update_bg = True if self.step_counter % self.n_step_bg == 0 else False
 
     def getNextTimeStep(self):
```
Comment (mcwimm) [inline TimeLoopLib/SimpleTimeLoop/SimpleLoop.py:36]: What do you think about adding some comments?

File: TimeLoopLib/SimpleTimeLoop/SimpleLoop.py [hunk 2]
```diff
@@ -27,22 +27,21 @@ def __init__(self, args):
...
-        t_2 = t_1 + self.delta_t
-        self.t_current = t_2
-        if (t_1 < self.t_end):
-            step_on = True
-            if (t_2 > self.t_end):
-                t_2 = self.t_end
-        elif (t_1 >= self.t_end):
-            step_on = False
+        self.getUpdateBools()
+        self.t_1 = self.t_current
```
Comment (mcwimm) [inline TimeLoopLib/SimpleTimeLoop/SimpleLoop.py:38]: Do we really need `self.t_current`?

replace ln 38 with:
`self.t_1 = self.t_2  # assigns t2 from previous timestep as new t1`
delete ln 40

**Review by mcwimm — COMMENTED** [review]
I've tested the code with an external time stepper script. The results are the same as with the previous implementation. However, the model crashes when all trees are dead, which did not happen before.
I guess the reason for the crash is that the AG-concept (AZOI) is called although no tree exist. And when calculating the max. height it crashes (can not calculate maximum of nothing).
I'm wondering how this was avoided before?

File: TimeLoopLib/SimpleTimeLoop/SimpleLoop.py (hunk 1 above)
Comment (jbathmann) [reply TimeLoopLib/SimpleTimeLoop/SimpleLoop.py:36]: ok..

File: TimeLoopLib/TreeDynamicTimeLoop.py [hunk 3]
```diff
@@ -14,23 +14,21 @@ def __init__(self, args):
         else:
             raise KeyError("Required time stepping not implemented.")
         print(case + " time stepping successfully initiated.")
-        self.step_on = True
 
     def iniSimpleTimeStepping(self, args):
         from .SimpleTimeLoop import SimpleLoop
         self.loop = SimpleLoop.Loop(args)
 
     def getNextTimeStepBoundaries(self):
```
[the remark this answers is not in the document]
Comment (jbathmann) [reply TimeLoopLib/TreeDynamicTimeLoop.py:22]: just a few...

File: TimeLoopLib/TreeDynamicTimeLoop.py [hunk 4]
```diff
@@ -14,23 +14,21 @@ def __init__(self, args):
...
-        while (self.step_on):
+        while (self.loop.step_on):
             print("Next time step to propagate" +
-                  " tree population with starting time " + str(self.t_ini) +
-                  " and end time " + str(self.t_end) + ".")
-            time_stepper.step(self.t_ini, self.t_end, self.update_ag,
-                              self.update_bg)
+                  " tree population with starting time " + str(self.loop.t_1) +
+                  " and end time " + str(self.loop.t_2) + ".")
+            time_stepper.step(self.loop.t_1, self.loop.t_2,
```
[the remark this answers is not in the document]
Comment (jbathmann) [reply TimeLoopLib/TreeDynamicTimeLoop.py:31]: Good idead!

File: TimeLoopLib/SimpleTimeLoop/SimpleLoop.py (hunk 2 above)
Comment (jbathmann) [reply TimeLoopLib/SimpleTimeLoop/SimpleLoop.py]: updated

**Comment by jbathmann** [comment]
> I've tested the code with an external time stepper script. The results are the same as with the previous implementation. However, the model crashes when all trees are dead, which did not happen before. I guess the reason for the crash is that the AG-concept (AZOI) is called although no tree exist. And when calculating the max. height it crashes (can not calculate maximum of nothing). I'm wondering how this was avoided before?

what exactly do you mean by "before"?

**Comment by mcwimm** [comment]
> > I've tested the code with an external time stepper script. The results are the same as with the previous implementation. However, the model crashes when all trees are dead, which did not happen before. I guess the reason for the crash is that the AG-concept (AZOI) is called although no tree exist. And when calculating the max. height it crashes (can not calculate maximum of nothing). I'm wondering how this was avoided before?
> 
> what exactly do you mean by "before"?

I was wrong. It never worked correctly.
The proposed changes in the step-function exit the MANGA execution if no more trees exist.
I've tested that with external and internal time stepping.

**Pull request merged**